src/include/ipc/header.h

```diff
@@ -19,18 +19,91 @@
+ *
+ * The SOF IPC is bi-directional. Messages can be initiated by the
+ * host and acknowledged by the DSP. Similarly, they can be initiated
+ * by the DSP and acknowledged by the host.
+ *
+ * IPC messages are divided into several groups: global, topology,
+ * power management, component, stream, DAI, trace, and a separate
+ * "firmware ready" message. Multiple messages can also be grouped
+ * into a message that belong to a compound group. Most messages are
+ * sent by the host to the DSP; only the following messages are sent
+ * by the DSP to the host:
+ *  - firmware ready: sent only once during initialization
+ *  - trace: optional, contains firmware trace data
+ *  - position update: only used if position data cannot be
+ *    transferred in a memory window or if forced by the kernel
+ *    configuration
+ *
+ * Message structure
+ * -----------------
+ *
+ * IPC messages have a fixed header and variable length payload.
+ *
+ *     +-------------------------------+
+ *     | IPC_HEADER (sof_ipc_cmd_hdr)  |
+ *     +-------------------------------+
+ *     | IPC_PAYLOAD (type specific)   |
+ *     +-------------------------------+
+ *
+ * The header contains size of the message and a 32bit command
+ * identifier with following structure:
  *
  * 0xGCCCNNNN where
  * - G is global cmd type (4 bits)
  * - C is command type (12 bits)
  * - N is the ID number (16 bits) - monotonic and overflows
  *
- * This is sent at the start of the IPM message in the mailbox. Messages should
- * not be sent in the doorbell (special exceptions for firmware).
-
+ * The command type define structure of the payload. E.g. for topology IPC
+ * message COMP_NEW structure is:
+ *
+ *     +-------------------------------------+
+ *     | IPC_HEADER (sof_ipc_cmd_hdr)        |
+ *     |    glb=GLB_TPLG_MSG                 |
+ *     |    type=COMP_NEW                    |
+ *     +-------------------------------------+
+ *     | IPC_PAYLOAD (sof_ipc_comp contents) |
+ *     +-------------------------------------+
+ *
+ * IPC ABI version compatibility rules
+ * -----------------------------------
+ *
+ * 1. FW binaries will only support one MAJOR ABI version which is advertised
+ *    to host at FW boot.
+ *
+ * 2. Host drivers will support the current and older MAJOR ABI versions of
+ *    the IPC ABI (up to a certain age to be determined by market information).
+ *
+ * 3. MINOR and PATCH ABI versions can differ between host and FW but must be
+ *    backwards compatible on both host and FW.
+ *
+ *    IPC messages sizes can be different for sender and receiver if MINOR or
+ *    PATCH ABI versions differ as new fields can be added to the end of
+ *    messages.
+ *
+ *    i) Sender > receiver: receiver only copies its own ABI structure size.
+ *
+ *    ii) Receiver > sender: receiver copies its own ABI size and zero pads
+ *                           new fields. i.e. new structure fields must be non
+ *                           zero to be activated.
```

Comment:
I got confused `Sender > receiver` because I wasn't sure whether it meant "sender version" or "sender size". Now I'm afraid it's confusing the two because it doesn't or didn't allow optional fields.

Anyway don't change this section yet so git can track it across the move.

Reply:
Ack, I'll add a note about this.